File: biggerStain.rb
#!/usr/bin/ruby
# frozen_string_literal:false

# esta clase contiene la logica para resolver el algoritmo de CCL mediante BFS
class StainFinder
  X_MOVE = [0, 0, 1, -1].freeze
  Y_MOVE = [1, -1, 0, 0].freeze

  def initialize(board)
    @input = board
    @height = board.length
    @width = board[0].length
    @visited = Array.new(@height) { Array.new(@width) { false } }
  end

  # Pixel class represent a poxel in the board
  class Pixel
    attr_reader :row, :column

    def initialize(row, column)
      @row = row
      @column = column
    end
  end

  def get_neighbours(pixel)
    my_neighbours = []
    0.upto(3) do |position|
      new_row = pixel.row + Y_MOVE[position]
      new_column = pixel.column + X_MOVE[position]
      if new_row < @height && new_row >= 0 && new_column >= 0 && new_column < @width
        my_neighbours.append(Pixel.new(new_row, new_column))
      end
    end
    my_neighbours
  end

  def are_the_same_color(pixel, other_pixel)
    @input[pixel.row][pixel.column] == @input[other_pixel.row][other_pixel.column]
  end

  def set_visited(pixel)
    @visited[pixel.row][pixel.column] = true
  end

  def is_visited(pixel)
    @visited[pixel.row][pixel.column]
  end

  def bfs(rootPixel)
    stainSize = 0
    remaining_pixels = [rootPixel]
    until remaining_pixels.empty?
      current_pixel = remaining_pixels.shift
      stainSize += 1
      get_neighbours(current_pixel).select do |neighbour_pixel|
        are_the_same_color(neighbour_pixel, current_pixel) and !is_visited(neighbour_pixel)
      end
                                   .each do |neighbour_pixel|
        set_visited(neighbour_pixel)
        remaining_pixels.append(neighbour_pixel)
      end
    end
    stainSize
  end

  # this is a bean class
  class Stain
    attr_accessor :size, :color

    def initialize
      self.size = 0
    end
  end

  def computeBiggerStain
    biggerStain = Stain.new
    @height.times do |row|
      @width.times do |column|
        next if @visited[row][column]

        @visited[row][column] = true
        stainSize = bfs(Pixel.new(row, column))
        if stainSize > biggerStain.size
          biggerStain.size = stainSize
          biggerStain.color = @input[row][column]
        end
      end
    end
    biggerStain
  end
end

def findMaxStainOnBoard(board, caseNumber)
  biggerStain = StainFinder.new(board).computeBiggerStain
  puts " #{caseNumber}. (\"#{biggerStain.color.chr}\",#{biggerStain.size})"
end

def main(pArgs)
  board = []
  new_case = 0
  File.readlines(pArgs).each do |line|
    if line[0] == '#'
      unless board.empty?
        findMaxStainOnBoard(board, new_case)
        board.clear
      end
      new_case = line.split(' ').last
    else
      board.append(line.bytes.tap(&:pop))
    end
  end
  findMaxStainOnBoard(board, new_case)
end

main(ARGV[0])

File: java/BiggerStain.java
import java.util.*;
import java.util.regex.Pattern;
import java.util.stream.IntStream;
import java.util.stream.Stream;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;

class BiggerStain {
    private static final short[] X_MOVE = { 0, 0, 1, -1 };
    private static final short[] Y_MOVE = { 1, -1, 0, 0 };
    private final int height;
    private final int width;
    private final byte[][] input;
    private final boolean[][] visited;

    public BiggerStain(byte[][] input) {
        this.input = input;
        height = input.length;
        width = input[0].length;
        visited = new boolean[height][width];
    }

    private static class Pixel {
        final int row;
        final int column;

        Pixel(int row, int column) {
            this.row = row;
            this.column = column;            
        }
        public int getRowIndex() {
            return row;
        }
        public int getColumnIndex() {
            return column;
        }
    }
    public List<Pixel> getNeighbours(Pixel pixel) {
        List<Pixel> myNeighbours = new ArrayList<>();
        IntStream.range(0, 4).forEach(position -> {
            int newRow = pixel.getRowIndex() + Y_MOVE[position];
            int newColumn = pixel.getColumnIndex() + X_MOVE[position];
            if (newRow < height && newRow >= 0 && newColumn >= 0 && newColumn < width){
                myNeighbours.add(new Pixel(newRow, newColumn));
            }
        });
        return myNeighbours;
    }
    
    public boolean areTheSameColor(Pixel pixel, Pixel otherPixel) {
        return input[pixel.getRowIndex()][pixel.getColumnIndex()] == 
                input[otherPixel.getRowIndex()][otherPixel.getColumnIndex()];
    }

    private void setVisited(Pixel pixel) {
        visited[pixel.getRowIndex()][pixel.getColumnIndex()] = true;
    }

    private boolean isVisited(Pixel pixel) {
        return visited[pixel.getRowIndex()][pixel.getColumnIndex()];
    }

    private int bfs(Pixel rootPixel) {
        int stainSize = 0;
        LinkedList<Pixel> remainingPixels = new LinkedList<>();
        remainingPixels.add(rootPixel);
        while (!remainingPixels.isEmpty()) {
            Pixel currentPixel = remainingPixels.poll();
            stainSize++;
            getNeighbours(currentPixel).stream().filter(
                    neighbourPixel -> areTheSameColor(neighbourPixel,currentPixel) && !isVisited(neighbourPixel))
                    .forEach(neighbourPixel -> {
                        setVisited(neighbourPixel);
                        remainingPixels.add(neighbourPixel);
                    });
        }
        return stainSize;
    }    
    private static class Stain{
        int color;
        int size;
        void setSize(int size){
            this.size=size;
        }
        void setColor(int color){
            this.color=color;
        }
        int getSize(){
            return size;
        }
        int getColor(){
            return color;
        }

    }
    Stain computeBiggerStain() {
        Stain biggerStain = new Stain();
        for (int row = 0; row < height; row++) {
            for (int column = 0; column < width; column++) {
                if (!visited[row][column]) {
                    int stainSize;
                    visited[row][column] = true;
                    if ((stainSize = bfs(new Pixel(row, column))) > biggerStain.getSize()) {
                        biggerStain.setSize(stainSize);
                        biggerStain.setColor(input[row][column]);
                    }
                }
            }
        }
        return biggerStain;
    }    

    private static void findMaxStainOnBoard(final List<byte[]> board, StringBuilder caseNumber) {
        byte[][] input = board.toArray(new byte[board.size()][board.get(0).length]);
        Stain biggerStain = new BiggerStain(input).computeBiggerStain();
        System.out.println(caseNumber + ".  (\"" + (char) biggerStain.getColor() + "\" , " + biggerStain.getSize() + ")");
    }

    public static void main(String[] pArgs) throws IOException {
        try (Stream<String> inputLines = Files.lines(Paths.get(pArgs[0]))) {
            final List<byte[]> board = new ArrayList<>();
            StringBuilder caseNumber = new StringBuilder();
            inputLines.forEach(line -> {
                if (line.charAt(0) == '#') {
                    if (!board.isEmpty()) {
                        findMaxStainOnBoard(board, caseNumber);
                        board.clear();
                        caseNumber.setLength(0);
                    }
                    String[] newCaseLine= Pattern.compile(" ").split(line);
                    caseNumber.append(newCaseLine[newCaseLine.length-1]);                    
                } else {
                    board.add(line.getBytes());
                }
            });
            findMaxStainOnBoard(board, caseNumber);
        }
    }
}
